src/Cryptol/TypeCheck/Unify.hs

```diff
@@ -111,21 +111,17 @@ bindVar v@(TVBound {}) t
   where k = kindOf v
 
 bindVar x@(TVFree _ _ xscope _) (TVar y@(TVFree _ _ yscope _))
-  | xscope `Set.isProperSubsetOf` yscope = return (singleSubst y (TVar x), [])
+  | xscope `Set.isProperSubsetOf` yscope =
+    return (uncheckedSingleSubst y (TVar x), [])
 
-bindVar x@(TVFree _ k inScope _d) t
+bindVar x@(TVFree _ k _scope _d) t
```

Comment:
Now that we don't need `_scope` it might be more readable to just remove this pattern, and change the first guard to `kindOf x /= kindOf t`

Reply:
Actually, even better, maybe we should move the kind check to the `singleSubst` function, as another invariant we have is that the kinds of the variables and their types match.